one/src/events.rs

```diff
@@ -48,17 +71,221 @@ async fn slurp(opts: SlurpOpts) -> Result<()> {
+                        .unwrap()
+                        .to_rfc3339_opts(SecondsFormat::Secs, true);
+                    println!(
+                        "TimeEvent({}) validated at Timestamp({}) = {}",
+                        opts.cid, timestamp, rfc3339
+                    )
+                }
+                Err(e) => println!("{} failed with error: {:?}", opts.cid, e),
+            }
+        } else {
+            if event.get_key("data").is_some() {
+                println!("UnsignedDataEvent({}) is not signed", opts.cid);
+            }
+            validate_data_event_payload(&opts.cid, &block_store, None).await?;
+        }
     }
     Ok(())
 }
 
+async fn validate_data_event_envelope(cid: &Cid, block_store: &SQLiteBlockStore) -> Result<String> {
+    let block = block_store.get(cid.to_owned()).await?.unwrap();
+    let envelope = CborValue::parse(&block)?;
+    let signatures_0 = envelope
+        .get_key("signatures")
+        .context("Not an envelope")?
+        .get_index(0)
+        .context("Not an envelope")?;
+    let protected = signatures_0
+        .get_key("protected")
+        .context("Not an envelope")?
+        .as_bytes()
+        .context("Not an envelope")?
+        .as_slice();
+    // Deserialize protected as JSON
+    let protected_json: serde_json::Value = serde_json::from_slice(protected)?;
+    let controller = protected_json.get("kid").unwrap().as_str().unwrap();
+    let protected: Base64UrlString = protected.into();
+    let signature: Base64UrlString = signatures_0
+        .get_key("signature")
+        .context("Not an envelope")?
+        .as_bytes()
+        .context("Not an envelope")?
+        .as_slice()
+        .into();
+    let payload: Base64UrlString = envelope
+        .get_key("payload")
+        .context("Not an envelope")?
+        .as_bytes()
+        .context("Not an envelope")?
+        .as_slice()
+        .into();
+    let compact = format!("{}.{}.{}", protected, payload, signature);
+    let did = DidDocument::new(controller);
+    let jwk = Jwk::new(&did).await.unwrap();
+    ssi::jws::decode_verify(&compact, &jwk)?;
+    validate_data_event_payload(cid, block_store, Some(did.id.clone())).await?;
+    Ok(did.id)
+}
+
+// TODO: Validate the Data Event payload structure
+// TODO: Validate CACAO
+async fn validate_data_event_payload(
+    _payload_cid: &Cid,
+    _block_store: &SQLiteBlockStore,
+    _controller: Option<String>,
+) -> Result<()> {
+    Ok(())
+}
+
+// To validate a Time Event, we need to prove:
+// - TimeEvent/prev == TimeEvent/proof/root/${TimeEvent/path}
+// - TimeEvent/proof/root is in the Root Store
+//
+// - If root not in local root store try to read it from txHash.
+// - Validated time is the time from the Root Store
+async fn validate_time_event(
+    cid: &Cid,
+    block_store: &SQLiteBlockStore,
+    root_store: &SQLiteRootStore,
+    eth_rpc: &impl EthRpc,
+) -> Result<i64> {
+    if let Some(block) = block_store.get(cid.to_owned()).await? {
```

Comment:
we already got the block from the block store in the caller, can we pass it in so we don't have to go back to the block store a second time?